onmt/translate/translator.py

```diff
@@ -1228,3 +1229,94 @@ def _score_target(
         gold_scores = gold_scores.sum(dim=0).view(-1)
 
         return gold_scores
+
+
+class ScoringPreparator(nn.Module):
+    import torch.nn as nn
+    from sacrebleu import corpus_bleu
+
+    def __init__(self, fields, opt):
+        self.fields = fields
+        self.opt = opt
+        import pyonmttok
+        self.tgt_tokenizer = pyonmttok.Tokenizer(
+            **self.opt.tgt_onmttok_kwargs)
+
+    def tokenize_batch(self, batch_side, side):
+        field = dict(self.fields)[side].base_field
+        tokenized_sentences = []
+        for i in range(batch_side.shape[1]):
+            tokens = []
+            for t in range(batch_side.shape[0]):
+                token = field.vocab.itos[batch_side[t, i, 0]]
+                if token == field.pad_token or token == field.eos_token:
+                    break
+                if token != field.init_token:
+                    tokens.append(token)
+            tokenized_sentences.append(tokens)
+        return tokenized_sentences
+
+    def build_sources_and_refs(self, batch, mode):
+        if mode == 'valid':
+            sources = []
+            refs = []
+            for example in batch.dataset.examples:
+                sources.append(example.src[0])
+                refs.append(example.tgt[0])
+        elif mode == 'train':
+            sources = self.tokenize_batch(batch.src[0], 'src')
+            refs = self.tokenize_batch(batch.tgt, 'tgt')
+        return sources, refs
+
+    def translate(self, model, batch, gpu_rank, mode):
+        from onmt.translate.translator import Translator
+        model_opt = self.opt
+        from onmt.utils.parse import ArgumentParser
+        parser = ArgumentParser()
+        onmt.opts.translate_opts(parser)
+        base_args = (["-model", "dummy"] + ["-src", "dummy"])
+        opt = parser.parse_args(base_args)
+        opt.gpu = gpu_rank
+        ArgumentParser.validate_translate_opts(opt)
+        ArgumentParser.update_model_opts(model_opt)
+        ArgumentParser.validate_model_opts(model_opt)
+        scorer = onmt.translate.GNMTGlobalScorer.from_opt(opt)
+        import os
+        import codecs
```

Comment:
Probably cleaner to put all the imports on consecutive lines at the begining of the method definition.